Allow specifying an error when failing back HTLC

alecchendev opened the pull request:
Still in progress. Fixes #1570. This PR adds `ChannelManager::fail_htlc_backwards_with_reason` to allow the user to input their own `HTLCFailReason` to send to peers when failing back.

### Notes
- Not sure how to handle comments/documentation for the new function, thought I saw somewhere else where there was a variant of a function that the description was mostly duplicated but I could be wrong.
- Would like to DRY up between `fail_htlc_backwards` and `fail_htlc_backwards_with_reason` but because the msat + height data is found individually for each HTLC to form the error (in `fail_htlc_backwards`), not sure how I would.

### To do
- Add test(s). I was planning to test the new function similarly to `fail_htlc_backwards`, however I couldn't really find where `fail_htlc_backwards` was explicitly tested, mainly just used/indirectly tested throughout `ln/functional_tests.rs`. Still figuring what the best way to test this is.

### Questions
- Could anyone explain (or point me to a resource that explains) what the failure codes in HTLCFailReason::Reason are? I couldn’t find much in the docs/comments, I saw some stuff about failure codes in BOLT#2 for malformed HTLCs but this seems more general than that?

This is my first PR on this repo so please forgive me if I'm asking something trivial! I'm going to try and participate in more code review to get more assimilated to the repo, but in the meantime if anyone is able to provide any feedback I'd greatly appreciate it.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3477,6 +3477,35 @@ where
+	/// [`ChannelManager::claim_funds`]), you should still monitor for
+	/// [`events::Event::PaymentClaimed`] events even for payments you intend to fail, especially on
+	/// startup during which time claims that were in-progress at shutdown may be replayed.
+	pub fn fail_htlc_backwards_with_reason(&self, payment_hash: &PaymentHash, reason: &HTLCFailReason) {
```
I'm confused why CI is passing on this PR, because `HTLCFailReason` is in a private module 🤔 

In any case, I think we should change this to be parameterized by a `failure_code: u16` and `failure_data: Vec<u8>` and revert the `onion_utils` changes

alecchendev replied:
Ah, makes sense!

valentinewallace reviewed (commented):
> Could anyone explain (or point me to a resource that explains) what the failure codes in HTLCFailReason::Reason are? 

Here are the possible failure codes/messages: https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages 

Re: testing, the `expect_payment_failed` macro in `functional_test_utils` checks a failed payment's error code and data :)

alecchendev commented:
Thanks so much for the feedback @valentinewallace! I'll try to address most of the suggestions in about an hour or so.

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3477,6 +3477,35 @@ where
+	/// While LDK will never claim a payment automatically on your behalf (i.e. without you calling
+	/// [`ChannelManager::claim_funds`]), you should still monitor for
+	/// [`events::Event::PaymentClaimed`] events even for payments you intend to fail, especially on
+	/// startup during which time claims that were in-progress at shutdown may be replayed.
```
Removed duplicate text and added link to BOLT 4, will wait on further context because on my own I'm not super sure about anything specific to add.

valentinewallace commented:
Looking pretty good to me. You'll want to amend your commit history at some point, we follow bitcoin core's guidelines: https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

valentinewallace commented:
Feel free to request a review once you've added tests or otherwise want another look

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3477,6 +3477,25 @@ where
+	/// 
+	/// See https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages for
+	/// valid failure codes and data.
+	pub fn fail_htlc_backwards_with_reason(&self, payment_hash: &PaymentHash, failure_code: u16, failure_data: Vec<u8>) {
```
After looking into the [failure messages in BOLT 4](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages) I was thinking that there might be cases where the user would want to fail back with an error message besides `incorrect_or_unknown_payment_details`, mainly `final_incorrect_cltv_expiry`, `final_incorrect_htlc_amount`, or a channel update as it mentions in the section about what a final node might error with (not sure if cases where these errors would occur are already handled somewhere else). My implementation doesn't handle these specific errors as of now; I was actually going to ask if that made sense. I was honestly not super sure about the specific situation where this might be needed, and was wondering if you had something in mind when opening #1570?

For passing in an arbitrary code and data, it seems the only way to create an `HTLCFailReason::Reason` at the moment is using an arbitrary code and data, at least from what I can tell [here](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/ln/onion_utils.rs#L643). These seem like they would be good to represent with enums, and then maybe expose that to the user?

alecchendev replied:
Sounds good. I'll try to get an implementation of this working sometime in the next 2 days.

alecchendev commented:
Will rebase tomorrow, but besides that I've mostly gotten through a new implementation that I think makes more sense as well as some tests.

I only implemented the four failure codes @TheBlueMatt mentioned (`unknown_payment_details`, `temporary_node_failure`, `required_channel_feature_missing`, `expiry_too_soon`) to start, but I still have to think about other ones I might want to add. Also might move the conversion from `FailureCode` to `HTLCFailReason` to a separate function especially if I add more.

In `fail_htlc_backwards_with_reason`, for the `FailureCode::ExpiryTooSoon` branch, I have a ton of `unwrap`s, not super sure what's the best way to handle these--I see other places that are doing similar things (getting a channel from a short channel id) take different approaches, e.g. they sometimes just unwrap, sometimes propagate an error, sometimes handle it based on local context. Seems like it would be sorta unreasonable to panic when a user just tries to fail back an HTLC, so I'm thinking I might want to default to a certain failure code or maybe propagate an error so they can choose how to handle it. I would need to take some more time to think about which places it makes sense to just unwrap vs. properly handle the error, so suggestions welcome.

Also had some miscellaneous questions:
- For the check commits CI check, should I try to be passing that all the time, or mainly only after squashing at the end?
- While working on this I mainly just wanted to run the one test, but it took quite a while to compile because I assume it compiles all the other tests even if it doesn't use them...is there a way around this?

alecchendev commented:
> Instead of merging in upstream changes, please rebase per the guide at https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#rebasing-changes

Oh I see, thanks.

> (and squash down fixup commits into one linear history that has logically separate commits that don't fix previous commits in the same PR).

Just to clarify, should I be squashing any time I make changes in the PR or just at the end?

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -291,6 +291,28 @@ struct ReceiveError {
+pub enum FailureCode {
+	/// General temporary failure of the processing node.
+	TemporaryNodeFailure             = 0x2000 | 2,
+	/// The processing node has a required feature which was not in this onion.
```
Just tried to reword and add some examples/elaboration here. Let me know if there's anything else I should include:)

valentinewallace replied:
> also maybe say "us" for "the processing node"

@TheBlueMatt I think using "you" instead of "us" would be more consistent with our other docs for these failure codes, will be reviewing accordingly

valentinewallace commented on lightning/src/util/events.rs:
```diff
@@ -491,10 +491,10 @@ pub enum Event {
-	/// [`ChannelManager::fail_htlc_backwards`] within the HTLC's timeout, the HTLC will be
+	/// [`ChannelManager::fail_htlc_backwards`] or [`ChannelManager::fail_htlc_backwards_with_reason`]
+	/// to free up resources for this HTLC and avoid network congestion.
+	/// If you fail to call either [`ChannelManager::claim_funds`],[`ChannelManager::fail_htlc_backwards`],
```
nit: 
```suggestion
	/// If you fail to call either [`ChannelManager::claim_funds`], [`ChannelManager::fail_htlc_backwards`],
```

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -291,6 +291,32 @@ struct ReceiveError {
+/// For more info on failure codes, see <https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages>.
+#[derive(Clone, Copy)]
+pub enum FailureCode {
+	/// A general temporary failure for otherwise unspecified errors that occur
+	/// for the entire node while handling the received onion. This can be used in situations
+	/// where other temporary error codes would not fit, while indicating to the origin node
+	/// it may want to retry the payment.
```
I think this could be a bit more concise:

```suggestion
	/// We had a temporary error processing the payment. Useful if no other error codes fit 
	/// and we want to indicate that the payer may want to retry.
```

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -291,6 +291,32 @@ struct ReceiveError {
+	/// it may want to retry the payment.
+	TemporaryNodeFailure             = 0x2000 | 2,
+	/// We have a required feature which was not in this onion.
+	/// You may wish to use this in cases where you require some additional metadata be provided
+	/// with the payment but which was not provided here.
```
```suggestion
	/// For example, we may require some additional metadata that was not provided with this payment.
```

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -291,6 +291,32 @@ struct ReceiveError {
+	/// You may wish to use this in cases where you require some additional metadata be provided
+	/// with the payment but which was not provided here.
+	RequiredNodeFeatureMissing       = 0x4000 | 0x2000 | 3,
+	/// The CLTV expiry is too close to the current block height for safe handling by us. This
+	/// should only be used when failing back as a forwarding node.
```
Note that `fail_with_reason` is only ever called in response to a `PaymentClaimable` event, so it'll never be called by a forwarding node 

```suggestion
	/// The CLTV expiry is too close to the current block height for safe handling. Useful if 
	/// the payer had an outdated view of your CLTV expiry requirements on send.
```
I'm not sure if this is exactly the right reason, but it would be good to provide an example of some use case for this error code

alecchendev replied:
> Note that fail_with_reason is only ever called in response to a PaymentClaimable event, so it'll never be called by a forwarding node

Ah I was thinking that, but wasn't sure if it might've been able to be used in a forwarding context, but that makes sense.

With that, in the spec it seems that final nodes are not supposed to specify `expiry_too_soon` when returning an error because differentiating this failure code (along with `incorrect_payment_amount`) allows for probing for the destination--for final nodes it merges all these failure codes into `incorrect_or_unknown_payment_details`. I'm wondering should this failure code be excluded then?

valentinewallace replied:
Ah that's a good point.. @TheBlueMatt can confirm, I think we shouldn't use failure codes that are only intended to be used by forwarding nodes?

alecchendev replied:
Makes sense, I'll remove this variant (sometime tomorrow morning)

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -291,6 +291,32 @@ struct ReceiveError {
+	/// required and not present, the `payment_secret` doesn't match the `payment_hash`, the
+	/// received amount is incorrect, the CLTV expiry of the HTLC is too close to the current
+	/// block height for safe handling, or `payment_metadata` isn't present when it should be.
+	/// This failure code should be used when acting as the final node in a route.
```
```suggestion
```
This error enum will only be used by final nodes :) Could also note that using this variant is equivalent to calling `fail_htlc_backwards`.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3472,16 +3498,45 @@ where
+						let short_channel_id = htlc.prev_hop.short_channel_id;
+						let (counterparty_node_id, chan_id) = self.short_to_chan_info.read().unwrap().get(&short_channel_id).cloned().unwrap();
+						let per_peer_state = self.per_peer_state.read().unwrap();
+						let peer_state = per_peer_state.get(&counterparty_node_id).unwrap().lock().unwrap();
+						let chan = peer_state.channel_by_id.get(&chan_id).unwrap();
```
Sadly neither of these `unwrap`s are safe.. I think we should default to the normal `IncorrectOrUnknownPaymentDetails` if either are `None`

alecchendev replied:
Just changed to handle the `unwrap`s from the `get`s. Just wanted to make sure that's what you meant right, and not also the `unwrap` after the `lock`?

alecchendev replied:
Oops my bad! Will fix those right now

alecchendev replied:
I think I got the other ones--if I've checked everything right the only unwraps should be from locks now.

Also would it be valuable for me test these specific paths where it's defaulting? And if so are there any existing tests (or general areas) you could point me to to see how I might simulate these situations?

alecchendev commented:
Btw I moved the code to fetch the corresponding error data to create the `HTLCFailReason` to a separate internal function `get_htlc_fail_reason_from_failure_code`. I rebased this to streamline the commits but realized leaving the separate commits temporarily would've probably been a helpful indicator, so just wanted to give a heads up.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -289,6 +289,27 @@ struct ReceiveError {
+	/// We have a required feature which was not in this onion. For example, you may require
+	/// some additional metadata that was not provided with this payment.
+	RequiredNodeFeatureMissing       = 0x4000 | 0x2000 | 3,
+	/// You may wish to use this when a `payment_hash` is unknown, the `payment_secret` is
+	/// required and not present, the `payment_secret` doesn't match the `payment_hash`, the
+	/// received amount is incorrect, the CLTV expiry of the HTLC is too close to the current
+	/// block height for safe handling, or `payment_metadata` isn't present when it should be.
```
Ah so some of these checks will be completed by LDK on the user's behalf:
- the `payment_secret` is required and not present
- `payment_secret` doesn't match the `payment_hash`
- received amount is incorrect
- `payment_metadata` isn't present when it should be 

Would suggest removing those

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -289,6 +289,27 @@ struct ReceiveError {
+	/// We have a required feature which was not in this onion. For example, you may require
+	/// some additional metadata that was not provided with this payment.
+	RequiredNodeFeatureMissing       = 0x4000 | 0x2000 | 3,
+	/// You may wish to use this when a `payment_hash` is unknown, the `payment_secret` is
```
```suggestion
	/// You may wish to use this when the `payment_preimage` is unknown, the `payment_secret` is
```

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3472,21 +3493,42 @@ where
 	}
 
+	/// Gets error data to form an [`HTLCFailReason`] given a [`FailureCode`] and [`ClaimableHTLC`].
+	/// If something goes wrong while finding the necessary error data, this defaults to getting
+	/// error data for [`FailureCode::IncorrectOrUnknownPaymentDetails`].
```
This is infallible now so these lines can be removed

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3472,16 +3496,45 @@ where
+	/// reason for the failure.
+	///
+	/// See [`FailureCode`] for valid failure codes.
+	pub fn fail_htlc_backwards_with_reason(&self, payment_hash: &PaymentHash, failure_code: &FailureCode) {
```
Oh huh, weird. Will do

alecchendev commented:
Okay, I think the commits should be all fixed up 👍

alecchendev commented:
Just squashed the first two commits together because the added function from the second commit was referenced in the first commit's documentation. *Now* it should be good I think.
